### PR TITLE
Dai functions

### PR DESCRIPTION
**Pull request by epheph**

### CODE REVIEW

File: contracts/DaiHrd.sol [hunk 1]
```diff
@@ -27,7 +28,7 @@ import { RuntimeConstants } from "./RuntimeConstants.sol";
  * are no special restrictions in the amount of tokens that created, moved, or
  * destroyed. This makes integration with ERC20 applications seamless.
  */
-contract ERC777 is RuntimeConstants, Context, IERC777, IERC20 {
+contract ERC777 is Context, IERC777, IERC20, RuntimeConstants {
```
Comment (MicahZoltu) [inline contracts/DaiHrd.sol]: Out of curiosity, why the change in order here?

File: contracts/DaiHrd.sol [hunk 2]
```diff
@@ -344,6 +345,7 @@ contract ERC777 is RuntimeConstants, Context, IERC777, IERC20 {
 		emit Transfer(address(0), account, amount);
 	}
 
+	// Note: Keydonix changed visibility from private to internal, we reference this function in derived contract
```
Comment (MicahZoltu) [inline contracts/DaiHrd.sol]: I feel like we should have a standardize comment style for these changes so they are easy to find (and thus re-apply) if we need to upgrade ERC777.  `// KEYDONIX: ...` maybe?

Comment (MicahZoltu) [reply contracts/DaiHrd.sol]: We should probably also add such a comment to the change we made to protect from the Uniswap bug.

File: contracts/DaiHrd.sol [hunk 3]
```diff
@@ -487,7 +490,57 @@ contract ERC777 is RuntimeConstants, Context, IERC777, IERC20 {
...
+
+	function updateAndFetchChi(Pot pot) internal returns (uint256 rontodaiPerPot) {
+		if (pot.rho() == now) {
+			rontodaiPerPot = pot.chi();
+		} else {
+			rontodaiPerPot = pot.drip();
+		}
+		return rontodaiPerPot;
+	}
+}
```
Comment (MicahZoltu) [inline contracts/DaiHrd.sol]: I'm assuming these are just copy-pasta from Maker?

Comment (MicahZoltu) [inline contracts/DaiHrd.sol]: I'm assuming these are just copy-pasta from Maker?

File: contracts/DaiHrd.sol [hunk 4]
```diff
@@ -526,4 +574,76 @@ contract DaiHrd is ERC777 {
 		vat.move(address(this), msg.sender, attorontodai);
 		return attorontodai;
 	}
+
+	// Dai specific functions. These functions all behave similar to standard functions
+	// with input or output denominated in Dai instead of DaiHrd
```
Comment (MicahZoltu) [inline contracts/DaiHrd.sol]: Style: no hard line breaks in prose text.

File: contracts/DaiHrd.sol (hunk 3 above)
Comment (epheph) [reply contracts/DaiHrd.sol]: one, rpow, rmul, mul are copies. The logic in calculatedChi is lifted.

But rontodaiPerPot is us, no one else uses that term. I should bring these chi ones back to the main contract, since they stopped looking like Maker code

File: contracts/DaiHrd.sol (hunk 1 above)
Comment (epheph) [reply contracts/DaiHrd.sol]: accident, i played around with changing hierarchy, put it back in the wrong order, i'll fix (although no-op, since there's no storage defined there)

File: contracts/DaiHrd.sol [hunk 5]
```diff
@@ -464,4 +504,99 @@ contract DaiHrd is ERC777 {
...
+		return convertAttodaiHrdToAttodai(attodaiHrd, rontodaiPerPot);
+	}
+
+	function totalSupplyDai() external view returns(uint256 attodai) {
+		uint256 rontodaiPerPot = calculatedChi();
+		attodai = convertAttodaiHrdToAttodai(totalSupply(), rontodaiPerPot);
+		return attodai;
+	}
+
+	function withdrawDai(uint256 attodai) external returns(uint256 attodaiHrd) {
```
Comment (MicahZoltu) [inline contracts/DaiHrd.sol]: Did we come to agreement on this function taking an `address`?  Argument for that change was that there is really no reason to withdraw DAI unless you are going to immediately be sending it away.

Comment (epheph) [reply contracts/DaiHrd.sol]: tell you what, if we add it, i want to check against 0x0, address(this), the dai address

Comment (MicahZoltu) [reply contracts/DaiHrd.sol]: The naming on this one is a bit weird since both withdraw functions withdraw DAI.  I wonder if we should rename `withdraw` to `withdrawDaiDenominatedInDaiHrd` and this one would just be `withdrawDai`?  Ugh.  Naming is hard.

Comment (MicahZoltu) [reply contracts/DaiHrd.sol]: Or, if we wanted consistency:
```
withdrawDai(attodaiHrd)
withdrawDaiDenominatedInDai(attodai)
```

Comment (MicahZoltu) [reply contracts/DaiHrd.sol]: I support `require(destination != address(0))` 💯

File: contracts/DaiHrd.sol [hunk 6]
```diff
@@ -464,4 +504,99 @@ contract DaiHrd is ERC777 {
...
+	function balanceOfDai(address tokenHolder) external view returns(uint256 attodai) {
+		uint256 rontodaiPerPot = calculatedChi();
+		uint256 attodaiHrd = balanceOf(tokenHolder);
+		return convertAttodaiHrdToAttodai(attodaiHrd, rontodaiPerPot);
+	}
+
+	function totalSupplyDai() external view returns(uint256 attodai) {
+		uint256 rontodaiPerPot = calculatedChi();
+		attodai = convertAttodaiHrdToAttodai(totalSupply(), rontodaiPerPot);
+		return attodai;
```
Comment (MicahZoltu) [inline contracts/DaiHrd.sol]: Style: In the function above you don't inline the `balanceOf` call but you do a direct return, in this one you inline the `totalSupply` call but then do a separate line return.  Is there a reason for the difference in style between these?

My preferenc in this case is probably a mix of the two
```solidity
return convertAttodaiHrdToAttodai(totalSupply(), rontodaiPerPot);
```

Comment (epheph) [reply contracts/DaiHrd.sol]: Trying to lean on maker's style, but i agree. i think the idea is "it's nice to label what you're returning" and then if you don't set it to that variable in the returns() , it feels like you're missing a step?

Comment (MicahZoltu) [reply contracts/DaiHrd.sol]: I really like named returns, it makes working with the ABI a lot better.  What I _dislike_ is the fact that you can accidentally not return and thus return an uninitialized variable.  I have lobbied in the past for Solc to have a hard requirement on there being a `return` statement present, even if you provide named returns but so far that hasn't happened.

For this project, I advocate for keeping the named return, but not for assigning to it and leveraging the implicit return value.  I'm fine with what you have in either of these as well though, my preference on the latter is very minor.

File: contracts/DaiHrd.sol [hunk 7]
```diff
@@ -464,4 +504,99 @@ contract DaiHrd is ERC777 {
...
+		attodaiHrd = attodai.mul(ONE).div(rontodaiPerPot) + 1;
+		return attodaiHrd;
+	}
+
+	function convertAttodaiHrdToAttodai(uint256 attodaiHrd, uint256 rontodaiPerPot ) internal pure returns (uint256 attodai) {
+		attodai = attodaiHrd.mul(rontodaiPerPot).div(ONE);
+		return attodai;
+	}
+
+	function calculatedChi() internal view returns (uint256 rontodaiPerPot) {
```
Comment (MicahZoltu) [inline contracts/DaiHrd.sol]: This could be `public` instead since it is just a `view` function.  May prove useful at some point since Maker doesn't have one.

File: contracts/DaiHrd.sol [hunk 8]
```diff
@@ -464,4 +504,99 @@ contract DaiHrd is ERC777 {
...
+		rontodaiPerPot = rmul(rpow(pot.dsr(), now - pot.rho(), ONE), pot.chi());
+		return rontodaiPerPot;
+	}
+
+	function updateAndFetchChi() internal returns (uint256 rontodaiPerPot) {
+		if (pot.rho() == now) {
+			rontodaiPerPot = pot.chi(); // replace with drop()?? TODO
```
Comment (MicahZoltu) [inline contracts/DaiHrd.sol]: It is rapidly looking less and less likely that `drop()` will make it in.  We may want to create a `public` helper method for:
```solidity
if (pot.rho() == now) return pot.drip();
else return pot.chi();
```
just so we don't have to duplicate that code (we do that in at least a few places).

Comment (MicahZoltu) [reply contracts/DaiHrd.sol]: Oh, 🤣 this _is_ that helper method!

File: contracts/DaiHrd.sol [hunk 9]
```diff
@@ -464,4 +504,99 @@ contract DaiHrd is ERC777 {
...
+		} else {
+			rontodaiPerPot = pot.drip();
+		}
+		return rontodaiPerPot;
+	}
+
+	// Takes whatever vat dai has already been transferred to DaiHrd, gives to pot (DSR) and mints tokens for user
+	function depositVatDaiForAccount(address account) internal returns (uint256 attopotDeposit) {
+		if (pot.rho() != now) pot.drip();
+		uint256 rontodaiPerPot = pot.chi();
```
Comment (MicahZoltu) [inline contracts/DaiHrd.sol]: Recommend changing this to use `updateAndFetchChi()`.
